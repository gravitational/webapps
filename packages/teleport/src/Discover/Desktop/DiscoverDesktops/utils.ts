import { Path } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/Path';
import { MonotoneX } from 'teleport/Discover/Desktop/DiscoverDesktops/lines/MonotoneX';

const STROKE_WIDTH = 4;

export interface Line {
  width: number;
  height: number;
  path: string;
}

export function createLine(
  desktopServiceElement: HTMLDivElement,
  desktopElement: HTMLDivElement,
  containerElement: HTMLDivElement
): Line {
  if (!desktopElement || !desktopServiceElement || !containerElement) {
    return null;
  }

  const desktopServiceRect = desktopServiceElement.getBoundingClientRect();
  const desktopRect = desktopElement.getBoundingClientRect();
  const containerRect = containerElement.getBoundingClientRect();

  const distance = desktopRect.left - desktopServiceRect.right;

  const path = new Path();
  const line = new MonotoneX(path);

  line.lineStart();

  const desktopLinePosition =
    desktopRect.top - containerRect.top + desktopRect.height / 2 - 1;
  const desktopServiceLinePosition =
    desktopServiceRect.top - containerRect.top + desktopServiceRect.height / 2;

  line.point(0, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(40, desktopServiceLinePosition - STROKE_WIDTH * 2);
  line.point(distance - 10, desktopLinePosition + STROKE_WIDTH / 2);
  line.point(distance, desktopLinePosition + STROKE_WIDTH / 2);

  line.lineEnd();

  return {
    width: distance,
    height: containerRect.height,
    path: path.path,
  };
}
