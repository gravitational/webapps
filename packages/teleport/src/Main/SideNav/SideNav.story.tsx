/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import * as Icons from 'design/Icon';
import { Box } from 'design';
import { ClusterSideNav } from './SideNav';

export default {
  title: 'TeleportCluster/SideNav',
};

export const SideNav = () => {
  const props = {
    ...defaultProps,
  };

  const inMemoryHistory = createMemoryHistory({});
  return (
    <Box
      mt={-3}
      height="100%"
      style={{ position: 'absolute', overflow: 'hidden' }}
    >
      <Router history={inMemoryHistory}>
        <ClusterSideNav {...props} />
      </Router>
    </Box>
  );
};

const clusterOptions = [
  { value: 'Kawupic', label: 'Kawupic' },
  { value: 'Ajiromil', label: 'Ajiromil' },
  { value: 'Wedolarav', label: 'Wedolarav' },
  { value: 'Urijorun', label: 'Urijorun' },
  { value: 'Irosutno', label: 'Irosutno' },
  { value: 'Wecdivrof', label: 'Wecdivrof' },
  { value: 'Nesbokti', label: 'Nesbokti' },
  { value: 'Nezfifku', label: 'Nezfifku' },
  { value: 'Cebfabca', label: 'Cebfabca' },
  { value: 'Cidbudud', label: 'Cidbudud' },
  { value: 'Jeakalu', label: 'Jeakalu' },
  { value: 'Luwcohba', label: 'Luwcohba' },
];

const defaultProps = {
  clusterOptions,
  clusterName: 'Wecdivrof',
  items: [
    {
      title: 'Apartment',
      Icon: Icons.Apartment,
      exact: true,
      to: '/web/site/apartment',
    },
    {
      title: 'Apple',
      Icon: Icons.Apple,
      exact: true,
      to: '/web/site/apple',
    },
    {
      title: 'Camera',
      Icon: Icons.Camera,
      exact: true,
      to: '/web/site/camera',
    },
  ],
  version: '6.0.0',
  homeUrl: '/localhost',
};
