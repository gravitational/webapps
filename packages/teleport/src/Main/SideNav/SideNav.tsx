/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import styled from 'styled-components';
import { NavLink } from 'react-router-dom';
import { Image, SideNav, SideNavItem } from 'design';
import SideNavItemIcon from 'design/SideNav/SideNavItemIcon';
import {
  useStoreNav,
  useStoreClusterId,
} from 'teleport/teleportContextProvider';
import logoSvg from './logo';
import cfg from 'teleport/config';

export default function Nav() {
  const clusterId = useStoreClusterId().getId();
  const items = useStoreNav().getSideItems();
  const $items = items.map((item, index) => (
    <SideNavItem
      key={index}
      as={NavLink}
      exact={item.exact}
      to={item.getLink(clusterId)}
    >
      <SideNavItemIcon as={item.Icon} />
      {item.title}
    </SideNavItem>
  ));

  return (
    <SideNav>
      <StyledLogoItem pl="4" width="208px" as="a" href={cfg.routes.app}>
        <Image src={logoSvg} mx="3" maxHeight="24px" maxWidth="160px" />
      </StyledLogoItem>
      <div
        style={{ display: 'flex', flexDirection: 'column', overflow: 'auto' }}
      >
        {$items}
      </div>
    </SideNav>
  );
}

const StyledLogoItem = styled(SideNavItem)`
  &:focus,
  &:active {
    border-left-color: transparent;
    background: inherit;
  }
`;
