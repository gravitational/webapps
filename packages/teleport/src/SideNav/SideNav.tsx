/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import styled from 'styled-components';
import { NavLink, Link } from 'react-router-dom';
import { Flex, Image } from 'design';

import cfg from 'teleport/config';

import SideNavItemIcon from './SideNavItemIcon';
import SideNavItem from './SideNavItem';
import SideNavItemGroup from './SideNavItemGroup';
import logoSvg from './logo';
import useSideNav from './useSideNav';

export default function Container() {
  const state = useSideNav();
  return <SideNav {...state} />;
}

export function SideNav(props: ReturnType<typeof useSideNav>) {
  const { items, path } = props;
  const $items = items.map((item, index) => {
    const isChild = item.items.length > 0;
    if (isChild) {
      return <SideNavItemGroup path={path} item={item} key={index} />;
    }

    return (
      <SideNavItem key={index} as={NavLink} exact={item.exact} to={item.route}>
        <SideNavItemIcon as={item.Icon} />
        {item.title}
      </SideNavItem>
    );
  });

  return (
    <Nav>
      <Logo />
      <Content>{$items}</Content>
    </Nav>
  );
}

export const Logo = () => (
  <LogoItem pl="4" width="208px" as={Link} to={cfg.routes.root}>
    <Image src={logoSvg} mx="3" maxHeight="24px" maxWidth="160px" />
  </LogoItem>
);

const LogoItem = styled(Flex)(
  props => `
  min-height: 56px;
  align-items: center;
  cursor: pointer;
  outline: none;
  text-decoration: none;
  width: 100%;
  &:hover {
    background ${props.theme.colors.primary.lighter};
    color ${props.theme.colors.primary.contrastText};
  }
`
);

export const Nav = styled.nav`
  background: ${props => props.theme.colors.primary.light};
  border-right: 1px solid ${props => props.theme.colors.primary.dark};
  min-width: 240px;
  width: 240px;
  overflow: auto;
  height: 100%;
  display: flex;
  flex-direction: column;
`;

export const Content = styled.div`
  display: flex;
  flex-direction: column;
  overflow: auto;
`;
