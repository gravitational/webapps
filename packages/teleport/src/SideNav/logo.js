import logoSvg from 'design/assets/images/teleport-logo.svg';
export default logoSvg;
