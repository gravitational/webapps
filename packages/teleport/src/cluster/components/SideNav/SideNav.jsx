/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { NavLink } from 'react-router-dom';
import { withState } from 'shared/hooks';
import { SideNav, SideNavItem } from 'design';
import SideNavItemIcon from 'design/SideNav/SideNavItemIcon';
import { useStoreUser, useStoreNav } from 'teleport/teleportContextProvider';
import cfg from 'teleport/config';

export function ClusterSideNav({ items }) {
  const $items = items.map((item, index) => (
    <SideNavItem key={index} as={NavLink} exact={item.exact} to={item.to}>
      <SideNavItemIcon as={item.Icon} />
      {item.title}
    </SideNavItem>
  ));

  return <SideNav>{$items}</SideNav>;
}

export default withState(() => {
  const items = useStoreNav().getSideItems();
  const { version } = useStoreUser().state;

  return {
    version,
    items,
    clusterName: cfg.clusterName,
    homeUrl: cfg.getClusterRoute(),
  };
})(ClusterSideNav);
