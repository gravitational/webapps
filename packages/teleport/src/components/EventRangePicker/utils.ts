import moment from 'moment';

export function getRangeOptions(): EventRange[] {
  return [
    {
      name: 'Today',
      from: moment(new Date())
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: '7 days',
      from: moment()
        .subtract(6, 'day')
        .startOf('day')
        .toDate(),
      to: moment(new Date())
        .endOf('day')
        .toDate(),
    },
    {
      name: 'Custom Range...',
      isCustom: true,
      from: new Date(),
      to: new Date(),
    },
  ];
}

export type EventRange = {
  from: Date;
  to: Date;
  isCustom?: boolean;
  name?: string;
};
