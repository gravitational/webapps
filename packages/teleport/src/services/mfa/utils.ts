/**
 * Copyright 2021 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Auth2faType, PreferredMfaType } from 'shared/services';

export type MfaOption = {
  value: Auth2faType;
  label: string;
};

export function getMfaOptions(
  mfa: Auth2faType,
  preferredMfa: PreferredMfaType,
  requireMfa = false
) {
  const mfaOptions: MfaOption[] = [];

  if (mfa === 'off' || !mfa) {
    return mfaOptions;
  }

  const mfaEnabled = mfa === 'on' || mfa === 'optional';
  const preferMfaWebauthn = mfaEnabled && preferredMfa === 'webauthn';
  const preferMfaU2f = mfaEnabled && preferredMfa === 'u2f';

  if (mfa === 'webauthn' || preferMfaWebauthn) {
    mfaOptions.push({ value: 'webauthn', label: 'Hardware Key' });
  }

  if (mfa === 'u2f' || preferMfaU2f) {
    mfaOptions.push({ value: 'u2f', label: 'Hardware Key' });
  }

  if (mfa === 'otp' || mfaEnabled) {
    mfaOptions.push({ value: 'otp', label: 'Authenticator App' });
  }

  if (!requireMfa && mfa === 'optional') {
    mfaOptions.push({ value: 'optional', label: 'None' });
  }

  return mfaOptions;
}
