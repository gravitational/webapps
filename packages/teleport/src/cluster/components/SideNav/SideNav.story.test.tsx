import React from 'react';
import { SideNav } from './SideNav.story';
import { render } from 'design/utils/testing';

test('rendering of SideNav', () => {
  const { container } = render(<SideNav />);
  expect(container.firstChild).toMatchSnapshot();
});
