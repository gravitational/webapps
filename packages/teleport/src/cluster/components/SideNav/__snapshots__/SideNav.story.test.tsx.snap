// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering of SideNav 1`] = `
.c4 {
  display: inline-block;
  transition: color .3s;
  margin-left: -40px;
  margin-right: 16px;
  color: #FFFFFF;
  font-size: 18px;
}

.c0 {
  box-sizing: border-box;
  margin-top: -16px;
  height: 100%;
}

.c1 {
  background: #222C59;
  border-right: 1px solid #111B48;
  min-width: 240px;
  width: 240px;
  overflow: auto;
  height: 100%;
  display: flex;
  flex-direction: column;
}

.c3 {
  box-sizing: border-box;
  padding-left: 64px;
  padding-right: 32px;
  color: rgba(255,255,255,0.87);
  background-color: #1C254D;
  display: flex;
  min-height: 56px;
  align-items: center;
  justify-content: flex-start;
  border-left: 4px solid transparent;
  cursor: pointer;
  outline: none;
  text-decoration: none;
  width: 100%;
  background: #222C59;
  color: rgba(255,255,255,0.56);
  font-size: 12px;
  font-weight: 600;
}

.c3:active,
.c3.active {
  border-left-color: #651FFF;
  background: #2C3A73;
  color: #FFFFFF;
}

.c3:hover,
.c3:focus {
  background: #2C3A73;
  color: #FFFFFF;
}

.c5 {
  display: inline-block;
  transition: color .3s;
  margin-left: -40px;
  margin-right: 16px;
  color: #FFFFFF;
  font-size: 18px;
  opacity: 0.56;
}

.c2:active .c5,
.c2.active .c5 {
  opacity: 1;
}

.c2:focus .c5,
.c2:hover .c5 {
  opacity: 1;
}

<div
  class="c0"
  height="100%"
  style="position: absolute; overflow: hidden;"
>
  <nav
    class="c1"
  >
    <a
      bg="primary.main"
      class="sc-fzqBZW c2 c3"
      color="text.primary"
      href="/web/site/apartment"
      pl="9"
      pr="5"
      theme="[object Object]"
    >
      <span
        class="c4 icon icon-apartment sc-AxjAm c5"
        color="light"
        font-size="4"
      />
      Apartment
    </a>
    <a
      bg="primary.main"
      class="sc-fzqBZW c2 c3"
      color="text.primary"
      href="/web/site/apple"
      pl="9"
      pr="5"
      theme="[object Object]"
    >
      <span
        class="c4 icon icon-apple sc-AxjAm c5"
        color="light"
        font-size="4"
      />
      Apple
    </a>
    <a
      bg="primary.main"
      class="sc-fzqBZW c2 c3"
      color="text.primary"
      href="/web/site/camera"
      pl="9"
      pr="5"
      theme="[object Object]"
    >
      <span
        class="c4 icon icon-camera sc-AxjAm c5"
        color="light"
        font-size="4"
      />
      Camera
    </a>
  </nav>
</div>
`;
